rfd/0126-backend-migrations.md

```diff
@@ -0,0 +1,503 @@
+time required to progress through each phase by increasing the Cloud release cadence.
+
+| Phase | Write                  | Read                                        | Details                                                                       |
+| ----- | ---------------------- | ------------------------------------------- | ----------------------------------------------------------------------------- |
+| 1     | Old key AND new key(s) | Old Key                                     | Introduce new resource(s) and mirror data to new key range on write           |
+| 2     | Old key AND new key(s) | New Key, fallback to old key if not present | Start reading from the new key and using the old key as a fall back           |
+| 3     | Old key AND new key(s) | New Key, fallback to old key if not present | Run data migration that copies all keys from old to new                       |
+| 4     | New key(s)             | New Key                                     | Stop mirroring writes and falling back to old resources when they don't exist |
+| 5     | New key(s)             | New Key                                     | Run optional migration that removes all old keys if they do not have a TTL    |
+
+#### Data Migration Phase
+
+Data migrations no longer need to occur block Auth initialization which will reduce the time it will take a new Auth
+instance to start processing requests. By waiting until Phase 3 to run data migrations when resource mirroring and
+fallback reads are already being performed by all Auth instances the migration can happen in the background.
+
+This phase is only required for resources which do not heartbeat. For any presence related resources we can use the
+write caused by the next heartbeat to perform the data migration.
+
+> The above only applies when doing phased migrations. Major releases used by self-hosted customers will still be using
+> the one-shot migration approach will still require running the migration during initialization.
+
+For example, `auth.Init` can discover if running in a Cloud environment via the modules package and launch all
+migrations in an async manner like so:
+
+```go
+func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
+  ...
+
+
+  if modules.GetModules().Features().Cloud {
+    go performMigrations()
+  } else {
+    performMigrations()
+  }
+
+  ...
+
+}
+
+```
+
+#### Feature Flags
+
+> Note: these should only be applied within Cloud where we have explicit control over the upgrade sequence
+
+Leveraging feature flags allows the code that handles the migrations to be decoupled from when the migrations actually
+occur. Flags should be defined and provided in a dynamic manner (i.e., environment variable/config option/cli flag) such
+that the flags can be toggled without having to publish a new version of Teleport.
+
+The following is an example backend service to persist `Foo` resources that uses feature flags to read/write new and/or
+old keys based on the flag.
+
```

Comment:
I think it would help to match the values of this hypothetical feature flag to the "phases" in the table above.

- `FooUseNewKeyOnly` is clearly phase 4.
- `FooUseOldKeyOnly` is not depicted in the table, I guess that would be pre-phase 0.
- `FooMirrorKeys` looks like phase 2.

How would we ever get into phase 1 with this example?